lib/plugin.js

```diff
@@ -2,93 +2,136 @@ var _ = require('lodash');
 var mongoose = require('mongoose');
 var common = require('./common');
 
-var currentVersion, migrations;
-var opts = {
-  collection: '',
-  migrations_path: process.cwd() + '/migrations/',
-  model: null,
-  version_field: '__m'
-}
-
 function Plugin(schema, options) {
-  var newField;
-
-  opts = _.merge(opts, options);
-  opts.collection = opts.collection.toLowerCase();
-
-  currentVersion = common.getCurrentVersion(opts.migrations_path, opts.collection);
-  migrations = common.listByVersion(opts.migrations_path, opts.collection);
-
-  // set VERSION_FIELD
-  newField = {};
-  newField[opts.version_field] = {
-    type: Number,
-    default: currentVersion
-  };
-
-  schema.add(newField);
-  schema.pre('init', preInit.bind(this));
-  schema.post('init', postInit.bind(this));
-};
-
-function preInit(next, data) {
-  data[opts.version_field] = data[opts.version_field] || 0;
-
-  // up
-  while (data[opts.version_field] < currentVersion) {
-    data = migrate(data, 'up');
-  }
-
-  // down
-  while (data[opts.version_field] > currentVersion) {
-    data = migrate(data, 'down');
-  }
-
-  next();
-}
-
-function migrate(data, action) {
-  var migration, migrationName, newData, unset;
-
-  if (action == 'up') {
-    data[opts.version_field]++;
-    migrationName = migrations[data[opts.version_field]];
-    migration = require(opts.migrations_path + migrationName).up;
-  } else if (action == 'down') {
-    migrationName = migrations[data[opts.version_field]];
-    migration = require(opts.migrations_path + migrationName).down;
-    data[opts.version_field]--;
-  }
-
-  if (!migration) {
-    console.error('mongoose-lazy-migrations', 'migration not found');
-    return data;
-  }
-
-  newData = migration(data);
-
-  if (!newData['$unset'] || typeof newData['$unset'] !== 'object') {
-    return newData;
-  }
-
-  for (var prop in newData['$unset']) {
-    delete newData[prop];
-  }
-
-  opts.model.update({ _id: newData._id }, { $unset: newData['$unset'] }, { strict: false }, errorHandler);
-
-  delete newData['$unset'];
-  return newData;
-}
-
-function postInit(data) {
-  data.update(data.toObject(), errorHandler);
-}
-
-function errorHandler(error) {
-  if (error) {
-    console.error('mongoose-lazy-migrations', error);
-  }
+    var newField;
+
+    var opts = _.merge({
+        collection: '',
+        migrations_path: process.cwd() + '/migrations/',
+        model: null,
+        version_field: '__m'
+    }, options);
+
+    // lowercase the collection
+    opts.collection = opts.collection.toLowerCase();
+
+    /**
+     * Helper function that gets the current version. We use this to dynamically get the version because
+     * caching it to a variable means we can't do multiple collections.
+     *
+     * @returns {number}
+     */
+    var getCurrentVersion = function() {
```

Comment:
@jaylensoeur this is the part I added. It is now used to "dynamically" get the current version rather than assigning it to a variable `currentVersion` when the plugin is instantiated.